### ISSUE TITLE
"Scale learning rate based on global batch size" and torch.cuda.Stream() in the main_amp.py example

### ISSUE THREAD

**Issue by vgoklani** [NVIDIA/apex#538]
Hey there,

First, thank you for posting this code example:

https://github.com/NVIDIA/apex/blob/master/examples/imagenet/main_amp.py

it was so incredibly helpful to see a fully worked example :)

I had one question about like 46:

    # Scale learning rate based on global batch size
    args.lr = args.lr*float(args.batch_size*args.world_size)/256. 

Why is there a static value of 256? More specifically, why 256? 

Also, why do you use a torch.cuda.Stream() in data_prefetcher()? Is this similar to the prefetch method in tf.data (from tensorflow) where you prefetch data before the iterator starts? Why do you need a stream?

Sorry if these questions are basic, I haven't seen them discussed before in a clearly explained example.

Thanks!

**Comment by mcarilli** [comment]
`main_amp.py` was originally based on the [official Pytorch imagenet example](https://github.com/pytorch/examples/blob/master/imagenet/main.py) which uses a [default batch size of 256](https://github.com/pytorch/examples/blob/master/imagenet/main.py#L39).  The default LR is also intended for a batch size of 256, so when I wrote the LR logic, it made sense to scale the LR relative to a base value of 256.

The Pytorch dataloader creates batches and applies augmentation purely on the CPU, so naturally each batch must be copied to the GPU before running a forward pass.  The prefetcher uses its own side stream so that prefetching (aka, [the CPU->GPU memcopy and normalization](https://github.com/NVIDIA/apex/blob/master/examples/imagenet/main_amp.py#L284-L299) of the _next_ iteration's data can overlap with the forward pass for the current iteration.  This means that every iteration can start immediately without having to wait for its batch to be copied from the CPU.

**Comment by vgoklani** [comment]
Thanks @mcarilli that was very helpful.
